drake/solvers/Constraint.h

```diff
@@ -182,17 +182,62 @@ class BoundingBoxConstraint : public LinearConstraint {
                          ub) {}
   virtual ~BoundingBoxConstraint() {}
 
-  virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x,
+  virtual void Eval(const Eigen::Ref<const Eigen::VectorXd>& x,
                     Eigen::VectorXd& y) const override {
-    y.resize(getNumConstraints());
+    y.resize(get_num_constraints());
     y = x;
   }
-  virtual void eval(const Eigen::Ref<const TaylorVecXd>& x,
+  virtual void Eval(const Eigen::Ref<const TaylorVecXd>& x,
                     TaylorVecXd& y) const override {
-    y.resize(getNumConstraints());
+    y.resize(get_num_constraints());
     y = x;
   }
 };
+
+
+/** LinearComplementarityConstraint
+ *
+ * Implements a constraint of the form:
+ *   Mx + q >= 0
+ *   x >= 0
+ *   x'(Mx + q) == 0
+ *
+ * An implied slack variable complements any 0 component of x, which
+ * (TODO ggould) will in the future be retrievable.
+ */
+class LinearComplementarityConstraint : public Constraint {
+ public:
+  template <typename DerivedM, typename Derivedq>
+  LinearComplementarityConstraint(const Eigen::MatrixBase<DerivedM>& M,
+                                  const Eigen::MatrixBase<Derivedq>& q)
+      : Constraint(q.rows(),
+                   Derivedq::Constant(0),
+                   Derivedq::Constant(
+                       std::numeric_limits<double>::infinity())),
+        m_matrix(M),
+        q(q) {}
+  virtual ~LinearComplementarityConstraint() {}
+
+  virtual void Eval(const Eigen::Ref<const Eigen::VectorXd>& x,
+                    Eigen::VectorXd& y) const override {
+    y.resize(get_num_constraints());
+    y = (m_matrix * x) + q;
+  }
+  virtual void Eval(const Eigen::Ref<const TaylorVecXd>& x,
+                    TaylorVecXd& y) const override {
+    y.resize(get_num_constraints());
+    y = (m_matrix.cast<TaylorVarXd>() * x) + q.cast<TaylorVarXd>();
+  };
+
+  const Eigen::MatrixXd& get_m_matrix() { return m_matrix; };
+  const Eigen::VectorXd& get_q() { return q; };
```

Comment:
Would be more consistent to call these `get_M()` and `get_q()`. I think that is appropriate here because it matches the math as documented for the class.